crates/iceberg/src/arrow/record_batch_transformer.rs

```diff
@@ -270,57 +394,199 @@ impl RecordBatchTransformer {
         snapshot_schema: &IcebergSchema,
         projected_iceberg_field_ids: &[i32],
         field_id_to_mapped_schema_map: HashMap<i32, (FieldRef, usize)>,
+        constants_map: HashMap<i32, PrimitiveLiteral>,
+        partition_spec: Option<&PartitionSpec>,
     ) -> Result<Vec<ColumnSource>> {
         let field_id_to_source_schema_map =
             Self::build_field_id_to_arrow_schema_map(source_schema)?;
 
-        projected_iceberg_field_ids.iter().map(|field_id|{
-            let (target_field, _) = field_id_to_mapped_schema_map.get(field_id).ok_or(
-                Error::new(ErrorKind::Unexpected, "could not find field in schema")
-            )?;
-            let target_type = target_field.data_type();
+        // Detect field ID conflicts that require name-based mapping.
+        //
+        // Conflicts occur when fields with initial_default (from add_files or schema evolution)
+        // have field IDs that match Parquet field IDs, but refer to different columns.
+        //
+        // Example from add_files (Hive table import):
+        //   Parquet file field IDs: name=1, subdept=2
+        //   Iceberg schema field IDs: id=1, name=2, dept=3, subdept=4
+        //   Partition columns (id, dept) have initial_default values
+        //
+        // Without name-based fallback, we'd incorrectly:
+        //   - Read partition column "id" (field_id=1) from Parquet field_id=1 (which is "name")
+        //   - Read data column "name" (field_id=2) from Parquet field_id=2 (which is "subdept")
+        //
+        // The fix: When conflicts exist, use name-based mapping for data columns.
+        // This matches Java's behavior in add_files procedure.
+        //
+        // Note: We do NOT treat identity-partitioned fields as conflicts. For identity partitions,
+        // it's NORMAL for the source column to exist in Parquet - we just use the constant from
+        // partition metadata instead of reading the file.
+        //
+        // See: TestAddFilesProcedure.addDataPartitionedByIdAndDept() in iceberg-java
+
+        // Build a set of source field IDs used in NON-identity partition transforms.
+        // These are regular data columns that happen to be used for partitioning (e.g., bucket, truncate).
+        // They should be read from Parquet files normally, not treated as partition columns.
+        let non_identity_partition_source_ids: std::collections::HashSet<i32> =
+            if let Some(spec) = partition_spec {
+                spec.fields()
+                    .iter()
+                    .filter(|f| !matches!(f.transform, Transform::Identity))
+                    .map(|f| f.source_id)
+                    .collect()
+            } else {
+                std::collections::HashSet::new()
+            };
+
+        let has_field_id_conflict = projected_iceberg_field_ids.iter().any(|field_id| {
```

Comment:
I'm confused about this part. I think the rule about missing column is quite clear in [iceberg spec](https://iceberg.apache.org/spec/#column-projection:~:text=Values%20for%20field%20ids%20which%20are%20not%20present%20in%20a%20data%20file%20must%20be%20resolved%20according%20the%20following%20rules%3A), it's just a fallback chain to lookup.